delta/rdf/src/main/scala/ch/epfl/bluebrain/nexus/delta/rdf/shacl/ShaclEngine.scala

```diff
@@ -0,0 +1,108 @@
+package ch.epfl.bluebrain.nexus.delta.rdf.shacl
+
+import java.net.URI
+import java.util
+
+import ch.epfl.bluebrain.nexus.delta.rdf.IriOrBNode.Iri
+import ch.epfl.bluebrain.nexus.delta.rdf.Vocabulary.nxsh
+import monix.bio.IO
+import org.apache.jena.query.Dataset
+import org.apache.jena.rdf.model._
+import org.topbraid.jenax.util.{ARQFactory, JenaDatatypes}
+import org.topbraid.shacl.arq.SHACLFunctions
+import org.topbraid.shacl.engine.{Constraint, ShapesGraph}
+import org.topbraid.shacl.util.{SHACLSystemModel, SHACLUtil}
+import org.topbraid.shacl.validation.{ValidationEngine, ValidationEngineConfiguration, ValidationUtil}
+
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Extend the [[ValidationEngine]] form TopQuadrant in order to add triples to the report
+  * with the number of targetedNodes
+  *
+  * @param dataset        the Dataset to operate on
+  * @param shapesGraphURI the URI of the shapes graph (must be in the dataset)
+  * @param shapesGraph    the ShapesGraph with the shapes to validate against
+  */
+// $COVERAGE-OFF$
+@SuppressWarnings(Array("NullParameter"))
+final class ShaclEngine private (dataset: Dataset, shapesGraphURI: URI, shapesGraph: ShapesGraph)
```

Comment:
this current API will not allow us to cache the shacl engine after the rules have been processed; does the lib allow for that?

Reply:
I don't understand your question. In any case, I will postpone anything related to the caching of schemas for when we are going to actually do it.

Reply:
The ShaclEngine takes both the shapes and the dataset to be validated during construction. This makes the engine instance dependent on the data itself. The problem with this is that the loading or "parsing" of the shacl rules would happen for each new resource, instead of happening only for changes in the shapes.
It's fine to defer this for when we'll work on the schema state.

Reply:
The validation of the shacl shapes does NOT happen every time, that's why we have the flag `validateShapes: Boolean` when validating. In the future we can make public `applySkipShapesCheck` and that's it.